File: tests/runners/csharp/tests/Test_02_imports.cs
// Test Case 02: Cross-namespace References
// Tests referencing types from different namespaces

using System;

class Program
{
    static int passed = 0;
    static int failed = 0;

    static void Assert(bool condition, string message)
    {
        if (!condition)
        {
            Console.WriteLine($"    FAILED: {message}");
            failed++;
        }
    }

    static void TestCommonEmbed()
    {
        Console.WriteLine("  Testing common embed (Position)...");

        var pos = new common.Position
        {
            x = 1.0f,
            y = 2.0f,
            z = 3.0f
        };

        Assert(pos.x == 1.0f, "x");
        Assert(pos.y == 2.0f, "y");
        Assert(pos.z == 3.0f, "z");

        Console.WriteLine("    PASS");
        passed++;
    }

    static void TestCommonEnum()
    {
        Console.WriteLine("  Testing common enum (Status)...");

        var status = common.Status.Active;
        Assert(status == common.Status.Active, "Active");

        status = common.Status.Inactive;
        Assert((int)status == 1, "Inactive value");

        Console.WriteLine("    PASS");
        passed++;
    }

    static void TestPlayerCrossNamespace()
    {
        Console.WriteLine("  Testing Player with cross-namespace types...");

        var player = new game.Player
        {
            id = 1,
            name = "Hero",
            position = new common.Position { x = 100.0f, y = 50.0f, z = 0.0f },
            status = common.Status.Active
        };

        Assert(player.id == 1, "id");
        Assert(player.name == "Hero", "name");
        Assert(player.position.x == 100.0f, "position.x");
        Assert(player.status == common.Status.Active, "status");

        Console.WriteLine("    PASS");
        passed++;
    }

    static void TestNpcCrossNamespace()
    {
        Console.WriteLine("  Testing NPC with cross-namespace types...");

        var npc = new game.NPC
        {
            id = 100,
            display_name = "Merchant",
            spawn_point = new common.Position { x = 50.0f, y = 50.0f, z = 0.0f },
            ai_state = common.Status.Active
        };

        Assert(npc.id == 100, "id");
        Assert(npc.display_name == "Merchant", "display_name");
        Assert(npc.spawn_point.x == 50.0f, "spawn_point.x");

        Console.WriteLine("    PASS");
        passed++;
    }

    static void TestBinaryCrossNamespace()
    {
        Console.WriteLine("  Testing binary serialization with cross-namespace types...");

        var original = new game.Player
        {
            id = 42,
            name = "Test Player",
            position = new common.Position { x = 123.456f, y = 789.012f, z = 345.678f },
            status = common.Status.Inactive
        };

        // Serialize
        using var ms = new System.IO.MemoryStream();
        using var writer = new System.IO.BinaryWriter(ms);
        game.BinaryWriters.WritePlayer(writer, original);

        // Deserialize
        ms.Position = 0;
        using var reader = new System.IO.BinaryReader(ms);
        var loaded = game.BinaryReaders.ReadPlayer(reader);

        Assert(loaded.id == original.id, "id mismatch");
        Assert(loaded.name == original.name, "name mismatch");
        Assert(Math.Abs(loaded.position.x - original.position.x) < 0.001f, "position.x mismatch");
        Assert(loaded.status == original.status, "status mismatch");

        Console.WriteLine($"    PASS (serialized {ms.Length} bytes)");
        passed++;
    }

    static void Main()
    {
        Console.WriteLine("=== Test Case 02: Cross-namespace References ===");

        TestCommonEmbed();
        TestCommonEnum();
        TestPlayerCrossNamespace();
        TestNpcCrossNamespace();
        TestBinaryCrossNamespace();

        if (failed > 0)
        {
            Console.WriteLine($"=== {failed} tests failed! ===");
            Environment.Exit(1);
        }
        else
        {
            Console.WriteLine("=== All tests passed! ===");
        }
    }
}

File: tests/runners/rust/tests/test_02_imports.rs
// Test Case 02: Cross-namespace References
// Tests referencing types from different namespaces

use std::io::Cursor;

use polygen_test::schema::common::{Position, Status};
use polygen_test::schema::game::{Player, NPC};
use polygen_test::schema_loaders::BinaryIO;

fn main() {
    println!("=== Test Case 02: Cross-namespace References ===");

    test_common_embed();
    test_common_enum();
    test_player_cross_namespace();
    test_npc_cross_namespace();
    test_binary_cross_namespace();

    println!("=== All tests passed! ===");
}

fn test_common_embed() {
    println!("  Testing common embed (Position)...");

    let pos = Position {
        x: 1.0,
        y: 2.0,
        z: 3.0,
    };

    assert_eq!(pos.x, 1.0);
    assert_eq!(pos.y, 2.0);
    assert_eq!(pos.z, 3.0);

    println!("    PASS");
}

fn test_common_enum() {
    println!("  Testing common enum (Status)...");

    let status = Status::Active;
    assert_eq!(status, Status::Active);

    let status = Status::Inactive;
    assert_eq!(status as i32, 1);

    println!("    PASS");
}

fn test_player_cross_namespace() {
    println!("  Testing Player with cross-namespace types...");

    let player = Player {
        id: 1,
        name: "Hero".to_string(),
        position: Position { x: 100.0, y: 50.0, z: 0.0 },
        status: Status::Active,
    };

    assert_eq!(player.id, 1);
    assert_eq!(player.name, "Hero");
    assert_eq!(player.position.x, 100.0);
    assert_eq!(player.status, Status::Active);

    println!("    PASS");
}

fn test_npc_cross_namespace() {
    println!("  Testing NPC with cross-namespace types...");

    let npc = NPC {
        id: 100,
        display_name: "Merchant".to_string(),
        spawn_point: Position { x: 50.0, y: 50.0, z: 0.0 },
        ai_state: Status::Active,
    };

    assert_eq!(npc.id, 100);
    assert_eq!(npc.display_name, "Merchant");
    assert_eq!(npc.spawn_point.x, 50.0);

    println!("    PASS");
}

fn test_binary_cross_namespace() {
    println!("  Testing binary serialization with cross-namespace types...");

    let original = Player {
        id: 42,
        name: "Test Player".to_string(),
        position: Position { x: 123.456, y: 789.012, z: 345.678 },
        status: Status::Inactive,
    };

    // Serialize
    let mut buffer = Vec::new();
    original.write_binary(&mut buffer).unwrap();

    // Deserialize
    let mut cursor = Cursor::new(&buffer);
    let loaded = Player::read_binary(&mut cursor).unwrap();

    assert_eq!(loaded.id, original.id);
    assert_eq!(loaded.name, original.name);
    assert!((loaded.position.x - original.position.x).abs() < 0.001);
    assert_eq!(loaded.status, original.status);

    println!("    PASS (serialized {} bytes)", buffer.len());
}
